# btp-module-chip-seq
Bioinformatics Training Platform (BTP) Module: ChIP-Seq

  * Topic
    * ChIP-Seq
  * Target Audience
    * Biologists
	* Non-bioinformaticians
	* Little to no programming expereience
  * Prerequisites
    * [btp-module-ngs-mapping](https://github.com/BPA-CSIRO-Workshops/btp-module-ngs-mapping)
  * Key Learning Outcomes
    *  Perform ChIP-Seq analysis, e.g. the detection of immuno-enriched areas using the chosen R package: ChIP-seq processing pipeline (SPP) 
  *  Visualize the peak regions through a genome browser, e.g. IGV or Ensembl, and identify the real peak regions
  *  Perform functional annotation using biomaRt R package and detect potential binding sites (motif) in the predicted binding regions using motif discovery tool, e.g. Trawler or MEME.

  * Time Required
    * 3.5 hrs

License
=======
The contents of this repository are released under the Creative Commons
Attribution 3.0 Unported License. For a summary of what this means,
please see:
http://creativecommons.org/licenses/by/3.0/deed.en_GB
